### PR TITLE
v0.24.2: Add DaiMigrator contract

### PR DESCRIPTION
**Pull request by BrendanChou**

### CODE REVIEW

File: contracts/external/traders/DaiMigrator.sol [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+            Require.that(
+                inputWei.isZero() && newInputPar.isZero(),
+                FILE,
+                "inputWei must be zero"
+            );
+        }
+
+        /* return the exact opposite amount of SAI in DAI */
+        return Types.AssetAmount ({
+            sign: !inputWei.sign,
```
Comment (AntonioJuliano) [inline contracts/external/traders/DaiMigrator.sol]: Ok so this is acting on the maker account, so it's negative as in it's giving away DAI. Makes sense 👍

File: contracts/external/traders/DaiMigrator.sol [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        Account.Info memory /* makerAccount */,
+        Account.Info memory takerAccount,
+        Types.Par memory oldInputPar,
+        Types.Par memory newInputPar,
+        Types.Wei memory inputWei,
+        bytes memory /* data */
+    )
+        public
+        returns (Types.AssetAmount memory)
+    {
```
[the remark this answers is not in the document]
Comment (BrendanChou) [reply contracts/external/traders/DaiMigrator.sol]: It should be view, changing it to that

File: contracts/external/traders/DaiMigrator.sol [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+            Require.that(
+                !newInputPar.isPositive(),
+                FILE,
+                "newInputPar cannot be positive"
+            );
+        } else {
+            Require.that(
+                inputWei.isZero() && newInputPar.isZero(),
+                FILE,
+                "inputWei must be zero"
```
[the remark this answers is not in the document]
Comment (BrendanChou) [reply contracts/external/traders/DaiMigrator.sol]: If we try to call target to zero migration on an account that has already been migrated for some reason.

File: contracts/external/traders/DaiMigrator.sol [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+                "newInputPar cannot be negative"
+            );
+        } else if (oldInputPar.isNegative()) {
+            Require.that(
+                inputWei.isPositive(),
+                FILE,
+                "inputWei must be positive"
+            );
+            Require.that(
+                !newInputPar.isPositive(),
```
[the remark this answers is not in the document]
Comment (BrendanChou) [reply contracts/external/traders/DaiMigrator.sol]: It is not, no

[the remark this answers is not in the document]
Comment (BrendanChou) [reply contracts/external/traders/DaiMigrator.sol]: Will add tests to illustrate that

File: contracts/external/traders/DaiMigrator.sol [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        public
+    {
+        for (uint256 i = 0; i < migrators.length; i++) {
+            g_migrators[migrators[i]] = true;
+        }
+    }
+
+    // ============ Admin Functions ============
+
+    function addMigrator(
```
[the remark this answers is not in the document]
Comment (BrendanChou) [reply contracts/external/traders/DaiMigrator.sol]: Figured it's safer to keep it just in case

**Pull request merged**